scipy/stats/_qmc.py

```diff
@@ -2045,15 +2058,18 @@ def _random(
             return np.empty((n, self.d))
 
         def in_limits(sample: np.ndarray) -> bool:
-            return (sample.max() <= 1.) and (sample.min() >= 0.)
+            for i in range(self.d):
+                if (sample[i] > self.u_bounds[i] or sample[i] < self.l_bounds[i]):
+                    return False
```

Comment:
We should be able to do this in a vectorised way to hopefully not get too much impact on speed.

Did you benchmark you changes? We actually don't have a benchmark for this function, maybe we can consider adding one.

Reply:
I have added the benchmark in my last commit and I tested it with the changes I think you suggested, replacing that section of the code with:
 ```python
 if np.any((sample > self.u_bounds) | (sample < self.l_bounds)):
     return False
```
However, this actually worsened its performance. What should I do?

Reply:
I would guess that it's a matter of size of the array. Do you have the numbers? I would expect that testing on a larger array would make the NumPy version faster. So I would use the NumPy version.

Reply:
Before:
```
=== ======== ============= ========== ========== ==========
--                                        n                
-------------------------- --------------------------------
d   radius   ncandidates      30        100        300    
=== ======== ============= ========== ========== ==========
1    0.2          30       3.13±0ms   2.39±0ms   2.73±0ms 
1    0.2          60       4.65±0ms   3.99±0ms   3.94±0ms 
1    0.2         120       6.04±0ms   6.01±0ms   5.92±0ms 
1    0.1          30       4.00±0ms   3.97±0ms   3.94±0ms 
1    0.1          60       6.71±0ms   6.69±0ms   6.67±0ms 
1    0.1         120       12.6±0ms   12.7±0ms   12.6±0ms 
1    0.05         30       7.72±0ms   7.17±0ms   7.32±0ms 
1    0.05         60       13.4±0ms   13.1±0ms   13.3±0ms 
1    0.05        120       29.5±0ms   25.2±0ms   25.0±0ms 
3    0.2          30       5.35±0ms   52.3±0ms   52.1±0ms 
3    0.2          60       6.20±0ms   99.9±0ms   104±0ms  
3    0.2         120       8.81±0ms   158±0ms    222±0ms  
3    0.1          30       4.69±0ms   19.6±0ms   85.1±0ms 
3    0.1          60       6.45±0ms   26.1±0ms   105±0ms  
3    0.1         120       9.11±0ms   37.0±0ms   195±0ms  
3    0.05         30       4.19±0ms   18.2±0ms   62.9±0ms 
3    0.05         60       5.59±0ms   24.4±0ms   86.4±0ms 
3    0.05        120       6.91±0ms   42.2±0ms   166±0ms  
5    0.2          30       6.37±0ms   19.1±0ms   61.8±0ms 
5    0.2          60       6.69±0ms   20.0±0ms   76.1±0ms 
5    0.2         120       7.29±0ms   26.4±0ms   98.6±0ms 
5    0.1          30       23.7±0ms   35.7±0ms   83.4±0ms 
5    0.1          60       23.4±0ms   40.4±0ms   95.4±0ms 
5    0.1         120       23.2±0ms   46.3±0ms   108±0ms  
5    0.05         30       487±0ms    503±0ms    566±0ms  
5    0.05         60       489±0ms    506±0ms    579±0ms  
5    0.05        120       488±0ms    508±0ms    581±0ms  
=== ======== ============= ========== ========== ==========

```
After:
```
=== ======== ============= ========== ========== ==========
--                                        n                
-------------------------- --------------------------------
d   radius   ncandidates      30        100        300    
=== ======== ============= ========== ========== ==========
1    0.2          30       3.05±0ms   3.04±0ms   2.99±0ms 
1    0.2          60       5.17±0ms   5.13±0ms   5.11±0ms 
1    0.2         120       8.16±0ms   8.20±0ms   8.13±0ms 
1    0.1          30       4.93±0ms   4.97±0ms   5.01±0ms 
1    0.1          60       8.32±0ms   8.40±0ms   8.35±0ms 
1    0.1         120       16.3±0ms   16.3±0ms   16.4±0ms 
1    0.05         30       8.84±0ms   8.88±0ms   8.91±0ms 
1    0.05         60       17.1±0ms   16.2±0ms   16.5±0ms 
1    0.05        120       32.6±0ms   32.3±0ms   32.9±0ms 
3    0.2          30       6.16±0ms   61.7±0ms   61.9±0ms 
3    0.2          60       7.12±0ms   121±0ms    128±0ms  
3    0.2         120       10.3±0ms   187±0ms    254±0ms  
3    0.1          30       5.41±0ms   22.0±0ms   95.9±0ms 
3    0.1          60       7.28±0ms   29.6±0ms   121±0ms  
3    0.1         120       10.2±0ms   42.1±0ms   221±0ms  
3    0.05         30       4.44±0ms   19.7±0ms   70.9±0ms 
3    0.05         60       6.10±0ms   25.7±0ms   98.2±0ms 
3    0.05        120       7.59±0ms   49.2±0ms   184±0ms  
5    0.2          30       6.27±0ms   20.1±0ms   66.0±0ms 
5    0.2          60       7.12±0ms   22.8±0ms   81.1±0ms 
5    0.2         120       7.63±0ms   28.1±0ms   104±0ms  
5    0.1          30       24.1±0ms   36.3±0ms   83.7±0ms 
5    0.1          60       23.4±0ms   41.3±0ms   99.7±0ms 
5    0.1         120       23.3±0ms   47.0±0ms   110±0ms  
5    0.05         30       489±0ms    501±0ms    548±0ms  
5    0.05         60       488±0ms    509±0ms    586±0ms  
5    0.05        120       484±0ms    516±0ms    582±0ms  
=== ======== ============= ========== ========== ==========
```